### DISCUSSION TITLE
Migrating to step-by-step migrations with existing project

### DISCUSSION THREAD

**Discussion by sqcsabbey** [simolus3/drift#3156] [Q&A] [answered]
I'm trying to migrate an existing project to use step-by-step migrations, and my first of these would be migrating from schema versions 22 to 23. When I run `drift_dev schema steps ...`, I get the following generated in _schema_versions.dart_:

```
import 'package:drift/internal/versioned_schema.dart' as i0;
import 'package:drift/drift.dart' as i1;
import 'package:drift/drift.dart'; // ignore_for_file: type=lint,unused_import

// GENERATED BY drift_dev, DO NOT MODIFY.
i0.MigrationStepWithVersion migrationSteps() {
  return (currentVersion, database) async {
    switch (currentVersion) {
      default:
        throw ArgumentError.value('Unknown migration from $currentVersion');
    }
  };
}

i1.OnUpgrade stepByStep() =>
    i0.VersionedSchema.stepByStepHelper(step: migrationSteps());
```

With this, when I try to call `migrationSteps`, the `from22To23` parameter isn't defined; this is how I tried to use it:

```
        await m.runMigrationSteps(
          from: max(22, from),
          to: to,
          steps: migrationSteps(
            from22To23: (m, schema) async {
              await m.addColumn(schema.features, schema.features.internal);
            },
          ),
        );
```

I've looked through the [documentation](https://drift.simonbinder.eu/docs/migrations/step_by_step/#generating-step-by-step-code), but I'm not sure what I'm missing. Could you please enlighten me?

**Comment by sqcsabbey** [comment, marked as answer]
My issue was that I tried to run `dart run drift_dev schema steps ...` against the dart versions of the frozen schemas, not the json ones. Do I have to still write the migration code (the `await m.addColumn(schema.features, schema.features.internal);` above), or is that handled now that the *schema_versions.dart* file generated successfully? The [documentation](https://drift.simonbinder.eu/docs/migrations/step_by_step/#generating-step-by-step-code) makes it look like this `addColumn` would be generated, but I don't see it in my *schema_versions.dart*.

**Reply by sqcsabbey** [reply]
Found where I need to implement `from22To23`, I think I'm all set with this (still need to try it)? 

I'm going to leave this here as someone else might run into a similar issue.

**Reply by simolus3** [reply]
> The [documentation](https://drift.simonbinder.eu/docs/migrations/step_by_step/#generating-step-by-step-code) makes it look like this `addColumn` would be generated, but I don't see it in my _schema_versions.dart_.

The docs mentions that the `migrationSteps` method will be generated and then shows how it would be used. So to clarify, you're still required to write your migrations manually. I tried figuring out migrations automatically by looking at diffs, but I think those approaches would always be unreliable since there is often more than one way to write migrations.
